File: Rian/fraction_organizer.py
import numpy as np
import sys

# Ler fractions.txt e extrair a diagonal
def read_diagonal(filename):
    data = np.loadtxt(filename)
    return np.diag(data)

# Ler fit_result.txt e extrair os nomes truncados
def read_truncated_names(filename):
    truncated_names = []
    with open(filename, 'r') as file:
        lines = file.readlines()
        for i in range(1, len(lines), 2):  # Pega a cada 2 linhas
            first_column = lines[i].split()[0]  # Pega a primeira coluna
            truncated_name = first_column.split('_')[0]  # Trunca no primeiro '_'
            truncated_names.append(truncated_name)
    return truncated_names

# Escrever arquivo de saída
def write_output(filename, names, diagonal_values):
    with open(filename, 'w') as file:
        for name, value in zip(names, diagonal_values):
            file.write(f"{name} {value}\n")

# Verifica argumentos da linha de comando
if len(sys.argv) != 2:
    print("Uso: python script.py <arquivo_de_saida>")
    sys.exit(1)

out_filename = sys.argv[1]

diagonal_elements = read_diagonal("/data1000/charmosinhos/Rian/Goofit3/GooFit/build/examples/D2PPK/Fit/"+out_filename+"/fractions.txt")
names = read_truncated_names("/data1000/charmosinhos/Rian/Goofit3/GooFit/build/examples/D2PPK/Fit/"+out_filename+"/fit_result.txt")
write_output("/data1000/charmosinhos/Rian/Goofit3/GooFit/build/examples/D2PPK/Fit/"+out_filename+"/diagonal_fractions.txt", names, 100*diagonal_elements)

print(f"Arquivo diagonal_fractions.txt para {out_filename} criado com sucesso!")

File: Rian/fraction_organizer.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

int fraction_organizer(const char* outfile="Laura_v1") {
    gSystem->Load("libRIO");
    gSystem->Load("libHist");
    std::string sfile1 = (std::string) "/data1000/charmosinhos/Rian/Goofit3/GooFit/build/examples/D2PPK/Fit/" + outfile + (std::string) "/fractions.txt";
    std::string sfile2 = (std::string) "/data1000/charmosinhos/Rian/Goofit3/GooFit/build/examples/D2PPK/Fit/" + outfile + (std::string) "/fit_result.txt";
    // Abrir o arquivo fractions.txt
    std::ifstream fractionsFile(sfile1.c_str());
    if (!fractionsFile) {
        std::cerr << "Erro ao abrir fractions.txt" << std::endl;
        return 1;
    }

    std::vector<double> diagonalElements;
    std::string line;
    int rowCount = 0;

    while (std::getline(fractionsFile, line)) {
        std::istringstream iss(line);
        double value;
        int column = 0;
        
        while (iss >> value) {
            if (column == rowCount) {
                diagonalElements.push_back(value);
            }
            column++;
        }
        rowCount++;
    }
    fractionsFile.close();

    // Abrir o arquivo fit_result.txt
    std::ifstream fitFile(sfile2.c_str());
    if (!fitFile) {
        std::cerr << "Erro ao abrir fit_result.txt" << std::endl;
        return 1;
    }

    std::vector<std::string> truncatedNames;
    int lineIndex = 0;
    while (std::getline(fitFile, line)) {
        if (lineIndex % 2 == 0) {  // Pegar a cada 2 linhas
            std::istringstream iss(line);
            std::string element;
            iss >> element; // Pega a primeira coluna

            size_t pos = element.find('_AMP');
            if (pos != std::string::npos) {
                element = element.substr(0, pos); // Truncar no primeiro '_'
            }
            truncatedNames.push_back(element);
        }
        lineIndex++;
    }
    fitFile.close();
  
    // Criar arquivo de saída diagonal_fractions.txt
  std::string sfile3 = (std::string) "/data1000/charmosinhos/Rian/Goofit3/GooFit/build/examples/D2PPK/Fit/" + outfile + (std::string) "/diagonal_fractions.txt";
    std::ofstream outFile(sfile3.c_str());
    if (!outFile) {
        std::cerr << "Erro ao criar diagonal_fractions.txt" << std::endl;
        return 1;
    }

    for (size_t i = 0; i < diagonalElements.size() && i < truncatedNames.size(); ++i) {
        outFile << truncatedNames[i] << " " << diagonalElements[i] << "\n";
    }

    outFile.close();
    std::cout << "Arquivo diagonal_fractions.txt criado com sucesso!" << std::endl;
    
    return 0;
}
